### PR TITLE
Provided support for Scraper Configuration.

### PR DESCRIPTION
**Pull request by M27afk**
With reference to https://github.com/ghostfolio/ghostfolio/issues/2292
@dtslvr 
Screenshot 2023-10-26 011351

### CODE REVIEW

File: apps/api/src/services/data-gathering/data-gathering.service.ts [hunk 1]
```diff
@@ -96,8 +97,9 @@ export class DataGatheringService {
         date
       );
 
-      const marketPrice =
-        historicalData[symbol][format(date, DATE_FORMAT)].marketPrice;
+      var marketPrice =
+        historicalData[symbol]?.[format(getYesterday(), DATE_FORMAT)]
```
Comment (dtslvr) [inline apps/api/src/services/data-gathering/data-gathering.service.ts:101]: This breaks the original logic, no? 🤔

Let's introduce a `isDryRun` param that you know you just need to return the latest value (without upsert)

File: apps/api/src/app/admin/admin.controller.ts [hunk 2]
```diff
@@ -221,7 +221,7 @@ export class AdminController {
     @Param('dataSource') dataSource: DataSource,
     @Param('dateString') dateString: string,
     @Param('symbol') symbol: string
```
Comment (dtslvr) [inline apps/api/src/app/admin/admin.controller.ts:223]: Add `@Query('dryRun') isDryRun?: boolean`

File: apps/api/src/app/admin/admin.controller.ts [hunk 3]
```diff
@@ -221,7 +221,7 @@ export class AdminController {
...
-  ): Promise<MarketData> {
+  ): Promise<MarketData | number> {
```
Comment (dtslvr) [inline apps/api/src/app/admin/admin.controller.ts:224]: Please keep `MarketData` only

File: apps/client/src/app/components/admin-market-data/asset-profile-dialog/asset-profile-dialog.component.ts [hunk 4]
```diff
@@ -240,6 +240,17 @@ export class AssetProfileDialog implements OnDestroy, OnInit {
       });
   }
 
+  public onTestScraper(symbol: string) {
+    const today = new Date();
+    this.adminService
+      .gatherSymbol({ dataSource: 'MANUAL', date: today, symbol })
+      .pipe(takeUntil(this.unsubscribeSubject))
+      .subscribe((response) => {
+        if (response !== null) alert('Current Market Price is: ' + response);
```
Comment (dtslvr) [inline apps/client/src/app/components/admin-market-data/asset-profile-dialog/asset-profile-dialog.component.ts:249]:
```
alert($localize`Current Market Price is:` + ' ' + response);
```



Can you use the explicit form with all the curly braces please for consistency?

File: apps/client/src/app/components/admin-market-data/asset-profile-dialog/asset-profile-dialog.component.ts [hunk 5]
```diff
@@ -240,6 +240,17 @@ export class AssetProfileDialog implements OnDestroy, OnInit {
       });
   }
 
+  public onTestScraper(symbol: string) {
+    const today = new Date();
+    this.adminService
+      .gatherSymbol({ dataSource: 'MANUAL', date: today, symbol })
+      .pipe(takeUntil(this.unsubscribeSubject))
+      .subscribe((response) => {
```
Comment (dtslvr) [inline apps/client/src/app/components/admin-market-data/asset-profile-dialog/asset-profile-dialog.component.ts:248]: `.subscribe(({ marketPrice }) => {`

File: apps/client/src/app/components/admin-market-data/asset-profile-dialog/asset-profile-dialog.component.ts [hunk 6]
```diff
@@ -240,6 +240,17 @@ export class AssetProfileDialog implements OnDestroy, OnInit {
...
+        if (response !== null) alert('Current Market Price is: ' + response);
+        else alert('Please try again.');
```
Comment (dtslvr) [inline apps/client/src/app/components/admin-market-data/asset-profile-dialog/asset-profile-dialog.component.ts:250]:
```
alert($localize`Please try again.`);
```

File: apps/client/src/app/services/admin.service.ts [hunk 7]
```diff
@@ -182,7 +182,7 @@ export class AdminService {
       url = `${url}/${format(date, DATE_FORMAT)}`;
     }
 
-    return this.http.post<MarketData | void>(url, {});
+    return this.http.post<MarketData | number | void>(url, {});
```
Comment (dtslvr) [inline apps/client/src/app/services/admin.service.ts:185]: Remove `number`

File: apps/client/src/app/components/admin-market-data/asset-profile-dialog/asset-profile-dialog.component.ts [hunk 8]
```diff
@@ -240,6 +240,17 @@ export class AssetProfileDialog implements OnDestroy, OnInit {
       });
   }
 
+  public onTestScraper(symbol: string) {
+    const today = new Date();
+    this.adminService
+      .gatherSymbol({ dataSource: 'MANUAL', date: today, symbol })
```
Comment (dtslvr) [inline apps/client/src/app/components/admin-market-data/asset-profile-dialog/asset-profile-dialog.component.ts:246]: Add `isDryRun: true`

**Review by dtslvr — CHANGES_REQUESTED** [review]
Hi @M27afk

Thank you very much for your contribution 🤓 I have got an idea to make the handling in the backend easier that it is a test run (`isDryRun`). Then you can keep the existing code mainly as is. What do you think?

File: apps/api/src/app/admin/admin.controller.ts (hunk 3 above)
Comment (M27afk) [reply apps/api/src/app/admin/admin.controller.ts:224]: Then how to return `marketPrice` from `data-gathering.service` which is of type `number` ?

File: apps/api/src/services/data-gathering/data-gathering.service.ts (hunk 1 above)
Comment (M27afk) [reply apps/api/src/services/data-gathering/data-gathering.service.ts:101]:
```
if(isDryRun)
      {return historicalData} or {return marketPrice} ?
```

I need help in maintaining return type `MarketData`

Comment (dtslvr) [reply apps/api/src/services/data-gathering/data-gathering.service.ts:101]: You can create a "synthetic" `MarketData` object, something like this:

```
if(isDryRun) {
  const date = new Date();
  
  ...

  return {
    dataSource,
    date,
    marketPrice,
    symbol,
    createdAt: null,
    id: null,
    state: 'INTRADAY'
  };
}
```

File: apps/api/src/app/admin/admin.controller.ts (hunk 3 above)
Comment (dtslvr) [reply apps/api/src/app/admin/admin.controller.ts:224]: See my comment [here](https://github.com/ghostfolio/ghostfolio/pull/2546#discussion_r1375308167).

File: apps/api/src/services/data-gathering/data-gathering.service.ts (hunk 1 above)
Comment (M27afk) [reply apps/api/src/services/data-gathering/data-gathering.service.ts:101]: This is working fine only if the `finally` block is removed. Somehow, even though the `isDryRun` is `true`, it was entering the `finally` block.

File: apps/client/src/app/components/admin-market-data/asset-profile-dialog/asset-profile-dialog.component.ts (hunk 5 above)
Comment (M27afk) [reply apps/client/src/app/components/admin-market-data/asset-profile-dialog/asset-profile-dialog.component.ts:248]: After providing the workaround https://github.com/ghostfolio/ghostfolio/pull/2546#discussion_r1375388256, type error is showing up, though the data is being returned.
Error:
`Property 'marketPrice' does not exist on type 'void | { createdAt: Date; dataSource: DataSource; date: Date; id: string; marketPrice: number; state: MarketDataState; symbol: string; }'.`

Data returned:
```
createdAt: null
dataSource: "MANUAL"
date: "2023-10-29T08:00:58.583Z"
id: null
marketPrice: 113.7639
state: "INTRADAY"
symbol: "trial2"
[[Prototype]]: Object
```

**Comment by dtslvr** [comment]
> Any updates?

I only see your initial commit. Do you have some improvements locally you did after my review?

**Comment by M27afk** [comment]
https://github.com/ghostfolio/ghostfolio/pull/2546#discussion_r1375388256

**Comment by dtslvr** [comment]
> I've tried to implement requested changes locally, and I've replied in comments.

I see. What about this [one](https://github.com/ghostfolio/ghostfolio/pull/2546#discussion_r1375212765)? I cannot merge it with a return type of `MarketData | number` (see [Robustness principle](https://en.m.wikipedia.org/wiki/Robustness_principle)).

**Comment by M27afk** [comment]
The return types are reset. The requested changes have been made, apart from https://github.com/ghostfolio/ghostfolio/pull/2546#discussion_r1375213020 due to type error thrown. Hence `response` is maintained as it is, @dtslvr it'd be of help if you can handle this.

**Comment by dtslvr** [comment]
Superseded by #2808.

**Pull request closed without merging**